clients/drpoints/drpoints.cpp

```diff
@@ -0,0 +1,439 @@
+static hashtable_t pc_to_id_map;
+
+// Global hash table to keep track of the execution count of BBs.
+// Key: unique BB ID, value: execution count in terms of BB instruction size.
+static hashtable_t bb_count_table;
+
+// Global unique BB counter used as ID. It must start from 1.
+static uint64_t unique_bb_count = 1;
+
+// Global instruction counter to keep track of when we reach the end of the user-defined
+// instruction interval.
+static uint64_t instr_count = 0;
+
+// List of Basic Block Vectors (BBVs).
+// This is a vector of vector pointers. Each element is a vector of pairs
+// <BB_ID, execution_count * BB_size> of type bb_id_count_pair_t.
+static drvector_t bbvs;
+
+// Keeps track of the number of threads of the application. We abort when we detect a
+// multi-threaded application as it's not supported yet.
+static int thread_count = 0;
+
+typedef struct _bb_id_count_pair_t {
+    uint64_t id;    // Derived from unique_bb_count.
+    uint64_t count; // execution_count * BB_size.
+} bb_id_count_pair_t;
+
+static void
+free_id_count_pair(void *entry)
+{
+    dr_global_free(entry, sizeof(bb_id_count_pair_t));
+}
+
+static void
+free_bbv(void *entry)
+{
+    drvector_t *vector = (drvector_t *)entry;
+    if (!drvector_delete(vector))
+        FATAL("ERROR: BBV drvector not deleted");
+    dr_global_free(vector, sizeof(drvector_t));
+}
+
+static void
+free_hit_count(void *entry)
+{
+    uint64_t *hit_count = (uint64_t *)entry;
+    dr_global_free(hit_count, sizeof(uint64_t));
+}
+
+static void
+add_to_bbv(void *key, void *payload, void *user_data)
+{
+    uint64_t count = *((uint64_t *)payload);
+    // Skip BBs that were not executed in the current instruction interval.
+    if (count == 0)
+        return;
+    uint64_t id = (uint64_t)(ptr_int_t)key;
+
+    bb_id_count_pair_t *id_count_pair =
+        (bb_id_count_pair_t *)dr_global_alloc(sizeof(bb_id_count_pair_t));
+    id_count_pair->id = id;
+    id_count_pair->count = count;
+
+    // Add BB frequency to BBV.
+    drvector_t *bbv = (drvector_t *)user_data;
+    drvector_append(bbv, id_count_pair);
+}
+
+// We add execution counters to the table at instrumentation time. We cannot remove them
+// from the hit_count_map when we reach the instruction interval at execution time, or the
+// next interval won't have a execution counter. So, we set them to zero.
+static void
+set_count_to_zero(void *payload)
+{
+    uint64_t *count = ((uint64_t *)payload);
+    *count = 0;
+}
+
+static void
+save_bbv()
+{
+    // Clear global instruction count.
+    // Currently we only support single-threaded applications, so we don't use any locking
+    // mechanism to set this global counter.
+    instr_count = 0;
+
+    // Save current bb_count_table (i.e., the BBV for the current instruction
+    //  interval).
+    drvector_t *bbv = (drvector_t *)dr_global_alloc(sizeof(drvector_t));
+    // We overshoot the initial size of the BBV vector to avoid resizing it.
```

Comment:
Not understanding "overshoot the initial size": aren't you setting the initial size to the precise size you need?